File: tools/Word Document Editor/scripts/docx_utils.py
"""
Shared utilities for Word document track changes operations.
"""

import zipfile
from xml.etree import ElementTree as ET
from datetime import datetime

# Word XML namespaces
NS = {
    'w': 'http://schemas.openxmlformats.org/wordprocessingml/2006/main',
    'w14': 'http://schemas.microsoft.com/office/word/2010/wordml',
}


def parse_date(date_str):
    """Parse Word date string to readable format."""
    if not date_str:
        return None
    try:
        # Word dates are typically ISO format or ticks
        if 'T' in date_str:
            dt = datetime.fromisoformat(date_str.replace('Z', '+00:00'))
            return dt.strftime('%Y-%m-%d %H:%M:%S')
        return date_str
    except:
        return date_str


def get_text_from_element(elem):
    """Extract text content from a Word XML element."""
    text_parts = []
    for t in elem.findall('.//w:t', NS):
        if t.text:
            text_parts.append(t.text)
    return ''.join(text_parts)


def list_track_changes(docx_path):
    """List all track changes in a Word document."""
    from pathlib import Path
    
    docx_path = Path(docx_path)
    
    if not docx_path.exists():
        raise FileNotFoundError(f"File not found: {docx_path}")
    
    if not docx_path.suffix.lower() == '.docx':
        raise ValueError("File must be a .docx file")
    
    changes = []
    change_id = 1
    
    with zipfile.ZipFile(docx_path, 'r') as docx:
        # Read document.xml
        try:
            doc_xml = docx.read('word/document.xml')
        except KeyError:
            raise ValueError("Invalid Word document structure")
        
        root = ET.fromstring(doc_xml)
        
        # Find all insertions (w:ins)
        for ins in root.findall('.//w:ins', NS):
            author = ins.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', 'Unknown')
            date = ins.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}date', '')
            text = get_text_from_element(ins)
            
            changes.append({
                'id': change_id,
                'type': 'Insertion',
                'author': author,
                'date': parse_date(date),
                'text': text,
                'status': 'Pending',
                'element': 'ins',
                'xml_element': ins
            })
            change_id += 1
        
        # Find all deletions (w:del)
        for dele in root.findall('.//w:del', NS):
            author = dele.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', 'Unknown')
            date = dele.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}date', '')
            text = get_text_from_element(dele)
            
            changes.append({
                'id': change_id,
                'type': 'Deletion',
                'author': author,
                'date': parse_date(date),
                'text': text,
                'status': 'Pending',
                'element': 'del',
                'xml_element': dele
            })
            change_id += 1
        
        # Find formatting changes (w:rPrChange, w:pPrChange)
        for fmt_change in root.findall('.//w:rPrChange', NS):
            author = fmt_change.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', 'Unknown')
            date = fmt_change.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}date', '')
            
            changes.append({
                'id': change_id,
                'type': 'Formatting (Run)',
                'author': author,
                'date': parse_date(date),
                'text': '[Formatting change]',
                'status': 'Pending',
                'element': 'rPrChange',
                'xml_element': fmt_change
            })
            change_id += 1
        
        for fmt_change in root.findall('.//w:pPrChange', NS):
            author = fmt_change.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', 'Unknown')
            date = fmt_change.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}date', '')
            
            changes.append({
                'id': change_id,
                'type': 'Formatting (Paragraph)',
                'author': author,
                'date': parse_date(date),
                'text': '[Formatting change]',
                'status': 'Pending',
                'element': 'pPrChange',
                'xml_element': fmt_change
            })
            change_id += 1
    
    return changes


def list_comments(docx_path):
    """List all comments in a Word document."""
    from pathlib import Path
    
    docx_path = Path(docx_path)
    
    if not docx_path.exists():
        raise FileNotFoundError(f"File not found: {docx_path}")
    
    if not docx_path.suffix.lower() == '.docx':
        raise ValueError("File must be a .docx file")
    
    comments = []
    comment_id = 1
    
    with zipfile.ZipFile(docx_path, 'r') as docx:
        # Check if comments.xml exists
        try:
            comments_xml = docx.read('word/comments.xml')
        except KeyError:
            # No comments file means no comments
            return comments
        
        comments_root = ET.fromstring(comments_xml)
        
        # Read document.xml to get comment ranges
        try:
            doc_xml = docx.read('word/document.xml')
        except KeyError:
            raise ValueError("Invalid Word document structure")
        
        doc_root = ET.fromstring(doc_xml)
        
        # Find all comment elements in comments.xml
        for comment_elem in comments_root.findall('.//w:comment', NS):
            comment_id_attr = comment_elem.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}id')
            author = comment_elem.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}author', 'Unknown')
            date = comment_elem.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}date', '')
            initials = comment_elem.get('{http://schemas.openxmlformats.org/wordprocessingml/2006/main}initials', '')
            
            # Get comment text
            text = get_text_from_element(comment_elem)
            
            # Try to find the commented text in document
            commented_text = ''
            for comment_ref in doc_root.findall(f'.//w:commentReference[@w:id="{comment_id_attr}"]', NS):
                # Find the paragraph containing this comment reference
                para = comment_ref
                while para is not None and not para.tag.endswith('}p'):
                    para = para.getparent()
                if para is not None:
                    # Get text from paragraph (simplified - gets all text in para)
                    commented_text = get_text_from_element(para)[:100]  # Limit length
            
            comments.append({
                'id': comment_id,
                'comment_id': comment_id_attr,
                'author': author,
                'initials': initials,
                'date': parse_date(date),
                'text': text,
                'commented_text': commented_text,
            })
            comment_id += 1
    
    return comments

File: tools/Word Document Editor/scripts/docx-utils.js
/**
 * Shared utilities for Word document track changes operations.
 */

import JSZip from 'jszip';
import { DOMParser, XMLSerializer } from '@xmldom/xmldom';
import xpath from 'xpath';
import { readFileSync, writeFileSync } from 'fs';

// Word XML namespace
const NS_W = 'http://schemas.openxmlformats.org/wordprocessingml/2006/main';

// XPath select with namespace
const select = xpath.useNamespaces({ w: NS_W });

/**
 * Parse Word date string to readable format
 */
export function parseDate(dateStr) {
  if (!dateStr) return null;
  try {
    if (dateStr.includes('T')) {
      const dt = new Date(dateStr);
      return dt.toISOString().slice(0, 19).replace('T', ' ');
    }
    return dateStr;
  } catch {
    return dateStr;
  }
}

/**
 * Extract text content from a Word XML element
 */
export function getTextFromElement(elem) {
  const textNodes = select('.//w:t/text()', elem);
  return textNodes.map(n => n.nodeValue || '').join('');
}

/**
 * Read a docx file and return JSZip instance
 */
export async function readDocx(docxPath) {
  const data = readFileSync(docxPath);
  return await JSZip.loadAsync(data);
}

/**
 * Write JSZip instance to docx file
 */
export async function writeDocx(zip, outputPath) {
  const content = await zip.generateAsync({ type: 'nodebuffer' });
  writeFileSync(outputPath, content);
}

/**
 * Parse XML string to document
 */
export function parseXml(xmlString) {
  const parser = new DOMParser();
  return parser.parseFromString(xmlString, 'application/xml');
}

/**
 * Serialize document to XML string
 */
export function serializeXml(doc) {
  const serializer = new XMLSerializer();
  return '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>' + serializer.serializeToString(doc);
}

/**
 * List all track changes in a Word document
 */
export async function listTrackChanges(docxPath) {
  const zip = await readDocx(docxPath);

  // Read document.xml
  const docXml = await zip.file('word/document.xml')?.async('string');
  if (!docXml) {
    throw new Error('Invalid Word document structure');
  }

  const doc = parseXml(docXml);
  const changes = [];
  let changeId = 1;

  // Find all insertions (w:ins)
  const insertions = select('//w:ins', doc);
  for (const ins of insertions) {
    const author = ins.getAttribute('w:author') || 'Unknown';
    const date = ins.getAttribute('w:date') || '';
    const text = getTextFromElement(ins);

    changes.push({
      id: changeId++,
      type: 'Insertion',
      author,
      date: parseDate(date),
      text,
      status: 'Pending',
      element: 'ins'
    });
  }

  // Find all deletions (w:del)
  const deletions = select('//w:del', doc);
  for (const del of deletions) {
    const author = del.getAttribute('w:author') || 'Unknown';
    const date = del.getAttribute('w:date') || '';
    const text = getTextFromElement(del);

    changes.push({
      id: changeId++,
      type: 'Deletion',
      author,
      date: parseDate(date),
      text,
      status: 'Pending',
      element: 'del'
    });
  }

  // Find formatting changes (w:rPrChange)
  const rPrChanges = select('//w:rPrChange', doc);
  for (const fmt of rPrChanges) {
    const author = fmt.getAttribute('w:author') || 'Unknown';
    const date = fmt.getAttribute('w:date') || '';

    changes.push({
      id: changeId++,
      type: 'Formatting (Run)',
      author,
      date: parseDate(date),
      text: '[Formatting change]',
      status: 'Pending',
      element: 'rPrChange'
    });
  }

  // Find paragraph formatting changes (w:pPrChange)
  const pPrChanges = select('//w:pPrChange', doc);
  for (const fmt of pPrChanges) {
    const author = fmt.getAttribute('w:author') || 'Unknown';
    const date = fmt.getAttribute('w:date') || '';

    changes.push({
      id: changeId++,
      type: 'Formatting (Paragraph)',
      author,
      date: parseDate(date),
      text: '[Formatting change]',
      status: 'Pending',
      element: 'pPrChange'
    });
  }

  return changes;
}

/**
 * List all comments in a Word document
 */
export async function listComments(docxPath) {
  const zip = await readDocx(docxPath);

  // Check if comments.xml exists
  const commentsXml = await zip.file('word/comments.xml')?.async('string');
  if (!commentsXml) {
    return []; // No comments
  }

  const commentsDoc = parseXml(commentsXml);
  const comments = [];
  let commentNum = 1;

  // Find all comment elements
  const commentElems = select('//w:comment', commentsDoc);
  for (const comment of commentElems) {
    const commentId = comment.getAttribute('w:id');
    const author = comment.getAttribute('w:author') || 'Unknown';
    const date = comment.getAttribute('w:date') || '';
    const initials = comment.getAttribute('w:initials') || '';
    const text = getTextFromElement(comment);

    comments.push({
      id: commentNum++,
      commentId,
      author,
      initials,
      date: parseDate(date),
      text,
      commentedText: '' // Would need to cross-reference with document.xml
    });
  }

  return comments;
}

export { select, NS_W };

